jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/Server.java

```diff
@@ -278,10 +278,21 @@ private Server(
 		try {
 			jmxConnection = pool.borrowObject(this);
 			ImmutableList.Builder<Result> results = ImmutableList.builder();
-			MBeanServerConnection connection = jmxConnection.getMBeanServerConnection();
 
-			for (ObjectName queryName : query.queryNames(connection)) {
-				results.addAll(query.fetchResults(connection, queryName));
+			if(query.getQueryType() == Query.QueryType.POLL) {
+				// Fetch results from mbeans
+				MBeanServerConnection connection = jmxConnection.getMBeanServerConnection();
+
+				for (ObjectName queryName : query.queryNames(connection)) {
+					results.addAll(query.fetchResults(connection, queryName));
+				}
+			} else {
+				// Subscribe to notifications
+				if(!query.isNotificationListenerRegistered(jmxConnection)) {
```

Comment:
A notification should probably not be executed, but registered at startup. The processing of results should be done when the notification is received, not polled. Basically, the notification listener should directly send results to the ResultProcessor...

Reply:
The design with polling was the simplest starting point I could think of. Just for my understanding, why do you think polling is not a good idea in this case?

Pushing (converted) notifications to registered ResultProcessors could also be implemented. 

I need to understand
- where I could register the notification listeners at startup
- how to retrieve the output writers for a given query

Are output writers thread safe?

Reply:
Polling an intermediate list of notification requires to keep that state for some time. As you noted in your comment, that state can grow unbounded and end in an OOM (so at least that queue should be bounded). Also, having a queue of notification will delay those notifications from being processed for a somewhat arbitrary reason. If in the end, we're functionally just doing polling of an attribute value, then why not do just that and use the current query mechanics.

Reply:
I agree to the points about polling. Whether the delay is acceptable depends on the use case I guess. For me having a few seconds delay is acceptable. I disagree that polling from a list of received notifications is equivalent to polling MBeans. In the former case, the publisher sends explicit events and we should receive all of them (if we stay connected). This is not guaranteed with polling of MBeans as we might miss intermediate attribute values.

One problem with the pushing design is that we need to re-register listeners when re-creating JMX connections. I guess MBeanServerConnectionFactory can't do this for us. On the other hand, it's the only one who knows when a connection is created. I was thinking that the initial listener registration could happen from JmxTransformer when we iterate over servers and queries. But how to handle broken and re-created connections is still a mistery to me. Would you have any idea?